src/protocol/sort/reshare.rs

```diff
@@ -0,0 +1,146 @@
+use axum::BoxError;
+
+use crate::{
+    field::Field,
+    helpers::{
+        mesh::{Gateway, Mesh},
+        prss::{PrssSpace, SpaceIndex},
+        Direction, Identity,
+    },
+    protocol::{DValue, RecordId, Step},
+    secret_sharing::Replicated,
+};
+
+/// Reshare(i, [x])
+// This implements reshare algorithm of "Efficient Secure Three-Party Sorting Protocol with an Honest Majority" at communication cost of 2R.
+// Input: Pi-1 and Pi+1 knows their secret shares
+// Output: At the end of the protocol, [all 3 helpers receive renewed shares of x with randomness added
+
+#[derive(Debug)]
+pub struct Reshare<'a, G, S> {
+    gateway: &'a G,
+    prss: &'a PrssSpace,
+    record_id: RecordId,
+    step: S,
+}
+
+impl<'a, G, S: Step + SpaceIndex> Reshare<'a, G, S> {
+    #[allow(dead_code)]
+    pub fn new(prss: &'a PrssSpace, gateway: &'a G, record_id: RecordId, step: S) -> Self {
+        Self {
+            gateway,
+            prss,
+            record_id,
+            step,
+        }
+    }
+
+    /// Steps
+    /// 1. While calculating for helper, we call PRSS to get randoms for helper who needs to receive reshares (say r0, r1)
+    ///    `to_helper.left` knows r0 (named r1) and `to_helper.right` knows r1 (named r0)
+    /// 2. `to_helper.left` calculates part1 = (input.0 + input.1) - r1
+    ///    `to_helper.right` calculates part2 = (input.0 - r0)
+    /// 3. `to_helper.left` and `to_helper.right` exchange their calculated shares
+    /// 4. Everyone sets their shares  
+    ///    `to_helper.left` = (r0, part1 + part2))
+    ///    `to_helper.right` = (part1 + part2, r1)
+    ///    `to_helper`      = (r1, r0)
+
+    #[allow(dead_code)]
+    pub async fn execute<F, M>(
+        self,
+        input: Replicated<F>,
+        to_helper: Identity,
+    ) -> Result<Replicated<F>, BoxError>
+    where
+        F: Field + 'static,
+        M: Mesh + 'a,
+        G: Gateway<M, S>,
+    {
+        let mut channel = self.gateway.get_channel(self.step);
+
+        let (r0, r1) = self.prss.generate_fields(self.record_id.into());
+
+        let inputs = input.as_tuple();
+        // `to_helper.left` calculates part1 = (input.0 + input.1) - r1 and sends part1 to `to_helper.right`
+        if channel.identity() == to_helper.peer(Direction::Left) {
+            let part1 = inputs.0 + inputs.1 - r1;
+            channel
+                .send(
+                    to_helper.peer(Direction::Right),
+                    self.record_id,
+                    DValue { d: part1 },
```

Comment:
What is the purpose of DValue here? is it for handling backpressure later on? Is it ok to reuse the same name or would it be better to choose a new name for each sub-protocol?

Reply:
`DValue` is multiplication-specific message type that indicates helper's intent to send its multiplication intermediate value to peer. We can invent another message type for reshare - that is better because allows us not to confuse two different intents (to multiply and to reshare)

Reply:
sounds good. Renaming this to PartValue